### PR TITLE
Fix missing performance metrics for VMI resources

### PR DESCRIPTION
**Pull request by machadovilaca**
### What this PR does
Before this PR:

`kubevirt_rest_client_requests_total` was missing datapoints for some resources such as VMIs

After this PR:

`kubevirt_rest_client_requests_total` has data for all tracked resources

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
Fix missing performance metrics for VMI resources
```

### CODE REVIEW

**Comment by machadovilaca** [comment]
@rthallisey @xpivarc

**Review by xpivarc — COMMENTED** [review]
/approve
/hold

**Comment by alaypatel07** [comment]
@machadovilaca @xpivarc what kind of tests can we write to makes sure refactoring metrics does not break perf/scale tests?

**Comment by machadovilaca** [comment]
> @machadovilaca @xpivarc what kind of tests can we write to makes sure refactoring metrics does not break perf/scale tests?

if we have a set of metrics and labels used by the perf/scale tests, we can create in the monitoring lane or so, a validation that those metrics exist and their value increased by a certain amount

**Comment by xpivarc** [comment]
@machadovilaca are you able to rebase?

**Comment by machadovilaca** [comment]
> @machadovilaca are you able to rebase?

updated @xpivarc

**Comment by alaypatel07** [comment]
https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/12096/pull-kubevirt-e2e-k8s-1.29-sig-performance/1806352571170820096/artifacts/performance/VMI-perf-audit-results.json

Looking at this audit file, all the metrics are present except GET-endpoints-count and UPDATE-endpoints-count

image
image


@xpivarc @machadovilaca can you check why those two metrics are missing in this PR's performance test run results?

**Comment by xpivarc** [comment]
@alaypatel07 This is expected. The endpoints were only used by leader election afaik. From https://github.com/kubevirt/kubevirt/pull/12097/files#diff-eb74f3825e06024b57cee95c2eee1eeea90b30d281a703cddeb4b25a0e287faeR42 we are not using the endpoints anymore but leases.

**Comment by xpivarc** [comment]
/cc @fossedihelm 
Federico can you have a quick look at the code for lgtm?

File: pkg/monitoring/metrics/common/client/metrics.go [hunk 1]
```diff
@@ -31,10 +31,12 @@ import (
 
 var resourceParsingRegexs []*regexp.Regexp
 
-func SetupMetrics() error {
+func RegisterRestConfigHooks() {
```
Comment (fossedihelm) [inline pkg/monitoring/metrics/common/client/metrics.go:34]: IIUC `RegisterRestConfigHooks` should always be executed before `GetKubevirtClientConfig`.
If this is true, maybe it is worth it to add a comment to document this expected behavior

**Review by fossedihelm — COMMENTED** [review]
@machadovilaca Code looks good. Just a decorator nit: I'd prefer to swap the commits, but it's absolutely *NOT* mandatory. 
Also, please look at the comment below.
Thank you!

File: pkg/monitoring/metrics/common/client/metrics.go (hunk 1 above)
Comment (machadovilaca) [reply pkg/monitoring/metrics/common/client/metrics.go:35]: added

**Comment by machadovilaca** [comment]
> @machadovilaca Code looks good. Just a decorator nit: I'd prefer to swap the commits, but it's absolutely _NOT_ mandatory. Also, please look at the comment below. Thank you!

@fossedihelm updated

**Comment by alaypatel07** [comment]
> @alaypatel07 This is expected. The endpoints were only used by leader election afaik. From https://github.com/kubevirt/kubevirt/pull/12097/files#diff-eb74f3825e06024b57cee95c2eee1eeea90b30d281a703cddeb4b25a0e287faeR42 we are not using the endpoints anymore but leases.

Interesting, in that case I wonder if we need to start these metrics for leases in sig-scale.

File: pkg/monitoring/metrics/common/client/metrics.go (hunk 1 above)
Comment (alaypatel07) [reply pkg/monitoring/metrics/common/client/metrics.go:35]: If the default is to add a hook, in the future releases can we make this part of the `GetKubevirtClientConfig` function?

For now this looks good

**Review by alaypatel07 — COMMENTED** [review]
/lgtm as well.

@xpivarc @rthallisey are we okay to cancel the hold on this?

File: pkg/monitoring/metrics/common/client/metrics.go (hunk 1 above)
Comment (enp0s3) [reply pkg/monitoring/metrics/common/client/metrics.go:35]: @machadovilaca Hi, can you please elaborate on the dependency between the two in the commit msg?

Comment (xpivarc) [reply pkg/monitoring/metrics/common/client/metrics.go:35]: @alaypatel07 There might be use cases where you don't want metrics, e.g tests...

**Comment by enp0s3** [comment]
@xpivarc Hi, IMO the code isn't self-explainable enough to reflect that dependency? when I will write a new controller that also related to VMI lifecycle and also uses leases, how would I know I need to call one func before another in order to maintain metrics stability wrt VMI?

**Comment by enp0s3** [comment]
@xpivarc would you remember this dep yourself when you will review future PRs?

File: pkg/monitoring/metrics/common/client/metrics.go (hunk 1 above)
Comment (enp0s3) [reply pkg/monitoring/metrics/common/client/metrics.go:35]: @xpivarc Then I should express whether I want metrics or not via the code i.e. via function initialization. How would I remember this dependency?

**Comment by alaypatel07** [comment]
@enp0s3 those are valid questions, I think we should file an issue for those so we can get to a better way of expressing that dependency.

File: pkg/monitoring/metrics/common/client/metrics.go (hunk 1 above)
Comment (xpivarc) [reply pkg/monitoring/metrics/common/client/metrics.go:35]: @enp0s3 that is possible. I believe this was not done as the code is following kubernetes client-go

**Comment by xpivarc** [comment]
> @xpivarc Hi, IMO the code isn't self-explainable enough to reflect that dependency. when I will write a new controller that also related to VMI lifecycle and also uses leases, how would I know I need to call one func before another in order to maintain metrics stability wrt VMI?

IMO one needs to know that they want metrics to be exposed. Then the person needs to investigate what kind of metrics are provided by client-go by default. The usage of `RegisterRestConfigHooks`  is up to anyone who we will write new controller and I think now the doc on the function explains when it needs to be called.

>@xpivarc would you remember this dep yourself when you will review future PRs?

Most likely yes.

**Comment by enp0s3** [comment]
@xpivarc Are you going to open an issue about it? 
@sradco I won't have BW to follow-up on this

**Comment by enp0s3** [comment]
> > @xpivarc Hi, IMO the code isn't self-explainable enough to reflect that dependency. when I will write a new controller that also related to VMI lifecycle and also uses leases, how would I know I need to call one func before another in order to maintain metrics stability wrt VMI?
> 
> IMO one needs to know that they want metrics to be exposed. Then the person needs to investigate what kind of metrics are provided by client-go by default. The usage of `RegisterRestConfigHooks` is up to anyone who we will write new controller and I think now the doc on the function explains when it needs to be called.
> 
I don't understand why one needs to do investigation instead of just reading the code and understand the dependencies.
> > @xpivarc would you remember this dep yourself when you will review future PRs?
> 
> Most likely yes.

**Pull request merged**